src/xtm_api.cc

```diff
@@ -0,0 +1,359 @@
+	 * to notify the producer thread about free space in queue.
+	 */
+	int producer_write_fd;
+	/**
+	 * Flags, that define queue behaviour.
+	 */
+	unsigned flags;
+	/**
+	 * Flag indicates, that producer couldn't put an item in
+	 * the queue, and is waiting for notification.
+	 */
+	bool is_producer_should_be_notified;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+static inline int
+notify_fd(int fd)
+{
+	static uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(fd, &tmp, sizeof(tmp))) < 0 && errno == EINTR)
+		;
+	return ((cnt >= 0 || errno == EAGAIN) ? 0 : -1);
+}
+
+static inline int
+create_fds(int *read_fd, int *write_fd)
+{
+	int fds[2];
+	assert(read_fd != NULL);
+	assert(write_fd != NULL);
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((fds[0] = eventfd(0, 0)) < 0)
+#else /* !defined(TARANTOOL_XTM_USE_EVENTFD) */
+	if (pipe(fds) < 0)
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+		return -1;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	*read_fd = *write_fd = fds[0];
+#else /* !defined(TARANTOOL_XTM_USE_EVENTFD) */
+	*read_fd = fds[0];
+	*write_fd = fds[1];
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+	return 0;
+}
+
+struct xtm_queue *
+xtm_queue_new(unsigned size, unsigned flags)
+{
+	int save_errno = 0;
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+	assert((flags & (~XTM_QUEUE_NEW_VALID_FLAGS)) == 0);
+	queue->flags = flags;
+	queue->is_producer_should_be_notified = false;
+
+	if (create_fds(&queue->consumer_read_fd,
+		       &queue->consumer_write_fd) < 0) {
+		save_errno = errno;
+		goto free_queue;
+	}
+
+	if (create_fds(&queue->producer_read_fd,
+		       &queue->producer_write_fd) < 0) {
+		save_errno = errno;
+		goto close_consumer_fds;
+	}
+
+	if (fcntl(queue->consumer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->consumer_write_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->producer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->producer_write_fd, F_SETFL, O_NONBLOCK) < 0) {
+		save_errno = errno;
+		goto close_producer_fds;
+	}
+	if (queue->queue.create(size) < 0) {
+		save_errno = EINVAL;
+		goto close_producer_fds;
+	}
+	return queue;
+
+close_producer_fds:
+	close(queue->producer_read_fd);
+	if (queue->producer_read_fd != queue->producer_write_fd)
+		close(queue->producer_write_fd);
+close_consumer_fds:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (((queue->flags & XTM_QUEUE_MUST_CLOSE_READFD) != 0) &&
```

Comment:
Since `queue->flags` are not used for regular pushes now, you could do the same as with the pushes - don't store the flags in the queue, and pass them to `xtm_queue_delete()` as a second argument. Would be consistent with push functions, and would save a few bytes in `struct xtm_queue`.

Reply:
Fixed